added alembic for database migrations

ibeex opened the pull request:
added support for alembic https://github.com/Pylons/pyramid-cookiecutter-alchemy/issues/5

Commit:
added alembic for database migrations

test only test* files

trying to clarify readme's

initalize_db update alembic to head

env.py use cookiecutter for import, changed file_template

rename

rename

import errors

file_templat also to production ini

trying to clarify readmes

migratios are optional

spacing

removed redundant

removed loadapp

stevepiercy commented on README.rst:
```diff
@@ -36,11 +36,16 @@ Usage
+    # Then you can use alembic to upgrade to that revision.
+    $ env/bin/alembic -c development.ini upgrade head
+    # Or initialize the database.
+    $ env/bin/initialize_{{ cookiecutter.repo_name }_db development.ini
```
I have a sinking feeling that enough users will just run through these steps in order, copy-pasting without first thinking or reading the entire stanza. It would be helpful to use formatting and language to inform the user. Here are suggestions:

```
3. Initialize the database. Read all the options first, then choose according to your needs.
```

```
    # Use alembic to generate revisions before you run initialize.
    $ env/bin/alembic -c development.ini revision --autogenerate -m "init"
    # Then you can use alembic either to (1) upgrade to that revision...
    $ env/bin/alembic -c development.ini upgrade head
    # ...or (2) initialize the database.
    $ env/bin/initialize_{{ cookiecutter.repo_name }}_db development.ini
```

Note there was a missing trailing `}`. I like this abstraction, because I recall a couple of times where beginners copy-pasted `env/bin/initialize_myproj_db development.ini` and wondered why it didn't work. It stands out with this change.

stevepiercy commented on hooks/post_gen_project.py:
```diff
@@ -45,7 +46,12 @@
+        %(alembic_cmd)s -c development.ini revision --autogenerate -m "init"
+        # Then you can use alembic to upgrade to that revision.
+        %(alembic_cmd)s -c development.ini upgrade head
+        # Or initialize the database.
```
See above comment and make consistent.

stevepiercy commented on {{cookiecutter.repo_name}}/README.txt:
```diff
@@ -20,10 +20,22 @@ Getting Started
+- Run database migrations.
+
+    env/bin/alembic -c development.ini ...
+
```
The instructions must be consistent throughout.

Also this is the only set of instructions where you indicate that the user can run migrations. Instructions should align with typical workflows.

ibeex commented:
fixed, readme's as suggested and made them consistent. Also added http://alembic.zzzcomputing.com/en/latest/cookbook.html#building-an-up-to-date-database-from-scratch to database init script

hyperknot commented on {{cookiecutter.repo_name}}/development.ini:
```diff
@@ -24,6 +24,10 @@ sqlalchemy.url = sqlite:///%(here)s/{{ cookiecutter.repo_name }}.sqlite
+[alembic]
+# path to migration scripts
+script_location = {{ cookiecutter.repo_name }}/alembic
+
```
I recommend adding:

```
file_template = %%(year)d_%%(month).2d_%%(day).2d_%%(hour).2d_%%(minute).2d_%%(second).2d_%%(rev)s
```

Or maybe a bit shorter version there, to have sortable revisions. The default hash value makes it impossible to find out what is the newest revision in a date-less system, like Git.

ibeex replied:
It is OK with me, is there maybe any other default value that can be changed? I wanted to put bare minimum in template rest left to user, maybe add this as commented?

hyperknot replied:
Up to you, but the default file naming is quite unusable in my opinion.

ibeex replied:
Changed to sort able added default to comment.

hyperknot commented:
I've replaced [inklesspen's old env.py](https://github.com/hyperknot/pyramid_alembic_mako/blob/master/pyramid_alembic/alembic/%2Bpackage%2B/alembic/env.py_tmpl) with this one's, and I'm receiving 

```
AttributeError: 'TransLogger' object has no attribute 'registry'
```

for line 13.

It is happening because I have request logging in development via:

```
from paste.translogger import TransLogger
format = '%(status)s %(REQUEST_METHOD)s %(REQUEST_URI)s'
app = TransLogger(app, format=format)
```

Is there a way now to make request logging without such hack? If not, can we make this script to detect the wrapped app?

ibeex commented:
I can replace env.py
`target_metadata = __import__(app.registry.__name__).models.meta.Base.metadata` with
`target_metadata = {{project}}.models.meta.Base.metadata` or somebody has any other idea?

hyperknot commented:
Looks good to me, with the added comment for clarity! I'd also add it to production.ini, just for clarity, as hopefully no one would like to add a revision in production.

ibeex commented:
@hyperknot  can you try now new env.py

hyperknot commented:
@ibeex works perfectly now. I did a manual replace for the `{{cookiecutter.project_name}}` so I cannot test that, but otherwise it's perfect.

stevepiercy commented on README.rst:
```diff
@@ -36,16 +36,18 @@ Usage
+    # Or initialize the database directly to latest revision.
+    $ env/bin/initialize_{{ cookiecutter.repo_name }}_db development.ini
+    # Later you can run database migrations.
+    $ env/bin/alembic -c development.ini ...
```
Now I see the complete picture. Sorry I didn't fully understand on my first pass. I have a few more revisions, and then we're good to go. This is less verbose and clearer. The changes should be applied to the other two files as well for consistency.

```
3. Initialize the database.

.. code-block:: bash

    # First use alembic to generate revisions.
    $ env/bin/alembic -c development.ini revision --autogenerate -m "init"
    # Then you can use alembic to either:
    # (a) upgrade to that revision:
    $ env/bin/alembic -c development.ini upgrade head
    # (b) or initialize the database directly to the latest revision:
    $ env/bin/initialize_{{ cookiecutter.repo_name }}_db development.ini
    # Later you can run database migrations.
    $ env/bin/alembic -c development.ini ...
```

ibeex replied:
added as suggested

hyperknot commented:
Sorry, as a long time alembic user the instructions are still not clear to me. I believe we should be recommending either alembic's init, or `Base.metadata.create_all(engine)` + stamping. 

Also, I am not sure what happens if we are stamping with "head" as "head" is not a revision hash.

All in all, since this is a starter template and people should be running it on empty projects, I recommend the following:
- init the db with alembic (rev + upgrade)
- use the initdb script only for adding a single entry to the database. That is, no stamping, no `Base.metadata.create_all(engine)`, etc. Just make it a reference implementation for a command line script accessing the database, so users can later reuse it for cron jobs, etc.

ibeex commented:
@hyperknot stamping with 'head' uses latest revision hash, I checked it. And yes you are right option a) and b) are 100% same thing when run for first time. But if later on user clears database  (or go to production) and there are already migrations a) will apply all migrations in order b) will create last state wont go step by step which can be off use if there are lots of migrations steps as stated here. http://alembic.zzzcomputing.com/en/latest/cookbook.html#building-an-up-to-date-database-from-scratch.

Maybe we can add third command only for reference, cron jobs and somehow clarify all this in README's @stevepiercy

stevepiercy commented:
I have some points for y'all to consider.

1. Cookiecutters are intended to get users started with a Pyramid project, not maintain one.
2. We do not want to incur more support and maintenance responsibility than absolutely necessary. Linking to the docs or guides of external projects is much preferred over trying to accommodate multiple use cases or best practices.
3. Much more complex projects can be created from our official cookiecutters, like https://github.com/mikeckennedy/cookiecutter-pyramid-talk-python-starter, because they are very basic.

hyperknot commented:
I agree, thus why I recommend just mentioning revision and upgrade and nothing else. I don't know how could I contribute to @ibeex's PR, but I'd try help as well.

ibeex commented:
Hi, regarding last comments I can try to implement this template so that it can be used with or without migrations, left it up to user. Readme would be like this. 

```
3. Initialize the database.
   
	
.. code-block:: bash

    # Important this template can be used with alembic for database migrations
  
    # (a) alembic version with migrations
    # First use alembic to generate revisions.
    $ env/bin/alembic -c development.ini revision --autogenerate -m "init"
    # upgrade to that revision:
    $ env/bin/alembic -c development.ini upgrade head
    # Later you can run database migrations.
    $ env/bin/alembic -c development.ini ...
    # Also there is commented code that can be used to 
    # initialize the database directly to the latest revision
    # {{cookiecutter.repo_name}}/scripts/initializedb.py

    # (b) standard version without migrations
    $ env/bin/initialize_myproj_db development.ini
```

stevepiercy commented:
I think this is a good improvement and we're on the right path.

I'm uncertain about this one command with the trailing `...`. We should be explicit about what those dots represent to avoid copy-pasta mistakes, either with an actual command or a `<man-style-replacement-indicator>`.

```bash
    $ env/bin/alembic -c development.ini ...
```

Otherwise I've made a few changes to set user expectations.

```rest
3. Initialize the database.

.. code-block:: bash

    # This cookiecutter provides support for initializing the database either
    # (a) with migrations through alembic, or
    # (b) without migrations.
    #
    # (a) Initialize the database with migrations through alembic:
    # First use alembic to generate revisions.
    $ env/bin/alembic -c development.ini revision --autogenerate -m "init"
    # upgrade to that revision:
    $ env/bin/alembic -c development.ini upgrade head
    # Later you can run database migrations.
    $ env/bin/alembic -c development.ini ...
    # Also there is commented code that can be used to
    # initialize the database directly to the latest revision.
    # {{cookiecutter.repo_name}}/scripts/initializedb.py
    #
    # (b) Initialize the database without migrations:
    $ env/bin/initialize_{{cookiecutter.repo_name}}_db development.ini
```

Thank you!

ibeex commented:
added changes as discussed

hyperknot commented:
I like the new approach! One problem though, the app / tutorial requires that silly `MyModel.name == 'one'` object so that the website is displayed.

What should we do about it?
- Either change the website not to require that MyModel object?
- Add initdb or a version of initdb _after_ alembic rev & alembic up?

My vote would be to do the first version, as that object is really just a dummy with no actual function in the view.

Is there a matching tutorial for this cookiecutter where the steps are described?

stevepiercy commented on README.rst:
```diff
@@ -36,11 +36,27 @@ Usage
+    # upgrade to that revision:
+    $ env/bin/alembic -c development.ini upgrade head
+    # Later you can run database migrations.
+    # Don't forget `-c development.ini` option.
```
This comment for "Don't forget..." is redundant to the commands above it. Please remove for all instructions.

stevepiercy commented:
@hyperknot that would be a potential separate issue/PR. This cookiecutter is used in a couple of official tutorials, where the wiki2 is primary:

* http://docs.pylonsproject.org/projects/pyramid/en/latest/tutorials/wiki2/installation.html#generate-a-pyramid-project-from-a-cookiecutter
* http://docs.pylonsproject.org/projects/pyramid/en/latest/quick_tour.html#databases

@mmerickel would have further information. I'd also like him to review this PR before merging.

mmerickel commented:
I have a lot of opinions on the sqla+alembic integration and I will need a bit of time to digest this but it looks like it's in a very good place from my quick skimming of the setup. I will try very hard to look at it over the weekend.

stevepiercy reviewed (approved):
Docs part looks good to me!

stevepiercy commented:
I will need to update the Pyramid docs according to the final approved version. I would appreciate not merging this PR immediately upon final approval to give me time to make those changes. Besides, I like pressing green buttons.

hyperknot commented:
Just an idea, what about making using migrations a choice in the cookiecutter setup? Something like:

- "Do you want to use migrations (alembic)?"

If not, we'd not even create the alembic specific bits.

It might make the cookiecutter templates a bit more complicated though.

stevepiercy commented:
@hyperknot that occurred to me, too. I think that once the dust settles for how to implement alembic, then we could consider offering it as a choice. Maybe a separate issue/PR?

hyperknot commented:
@stevepiercy yes, I agree it should be a separate issue.

Any conclusion about how to go about the dummy object?

ibeex commented:
@hyperknot funny when I try new project with alembic db generation it works normally. We could remove check for dummy object and left only check if there is db initialized empty or not.

hyperknot commented:
@ibeex I agree that we should be just checking for the table being initialized, not the dummy object. But it might need changing some bits in the tutorial.

hyperknot commented:
@ibeex there are some problems with the current env.py.

Normally it's used to set up the database, so that the app can be started.
However, right now, it seems to be loading and running the full main() function of the app for the `command.stamp()` line.

```
Traceback (most recent call last):
  File "/shared/init_db.py", line 49, in <module>
    command.stamp(alembic_cfg, 'head')
  File "site-packages/alembic/command.py", line 496, in stamp
    script.run_env()
  File "site-packages/alembic/script/base.py", line 416, in run_env
    util.load_python_file(self.dir, 'env.py')
  File "site-packages/alembic/util/pyfiles.py", line 93, in load_python_file
    module = load_module_py(module_id, path)
  File "site-packages/alembic/util/compat.py", line 75, in load_module_py
    mod = imp.load_source(module_id, path, fp)
  File "alembic/env.py", line 13, in <module>
    app = loadapp('config:%s' % config.config_file_name, relative_to='.')
  File "site-packages/paste/deploy/loadwsgi.py", line 247, in loadapp
    return loadobj(APP, uri, name=name, **kw)
  File "site-packages/paste/deploy/loadwsgi.py", line 272, in loadobj
    return context.create()
  File "site-packages/paste/deploy/loadwsgi.py", line 710, in create
    return self.object_type.invoke(self)
  File "site-packages/paste/deploy/loadwsgi.py", line 146, in invoke
    return fix_call(context.object, context.global_conf, **context.local_conf)
  File "site-packages/paste/deploy/util.py", line 55, in fix_call
    val = callable(*args, **kw)
  File "__init__.py", line 76, in main
    if config_data['production'] and get_siteconfig(dbsession)['sentry_key_backend']:
  File "utils/siteconfig.py", line 36, in get_siteconfig
    data = JSONBlob.get_by_slug(dbsession, 'siteconfig').data_
AttributeError: 'NoneType' object has no attribute 'data_'
```
The problematic line is `File "__init__.py", line 76, in main` which is run for some reason.

hyperknot commented:
I think it's loadapp which might be trying to load the full app. Actually, what was the problem with the old approach? https://github.com/hyperknot/pyramid_alembic_mako/blob/master/pyramid_alembic/alembic/%2Bpackage%2B/alembic/env.py_tmpl

ibeex commented:
@hyperknot can you try now, I removed app part, tried new template with alembic, and it works.

hyperknot commented:
@ibeex the new code works nicely!

stevepiercy commented:
@ibeex would you please rebase your branch and push?

@mmerickel would you please review and comment?

ibeex commented:
@stevepiercy  I did rebase and push

hyperknot commented:
What is the status of this? I'd like to deprecate pyramid_alembic_mako and redirect to this, if it gets merged.

ibeex commented:
@hyperknot I don't know what is latest status, probably in some limbo :smile:

hyperknot commented:
It looks like everything was approved except the merge didn't happen.

stevepiercy commented:
@mmerickel expressed that he wanted to review this PR.

mmerickel commented on {{cookiecutter.repo_name}}/setup.py:
```diff
@@ -19,6 +19,7 @@
     'transaction',
     'zope.sqlalchemy',
     'waitress',
+    'alembic',
```
alphabetical

mmerickel commented on {{cookiecutter.repo_name}}/production.ini:
```diff
@@ -20,6 +20,12 @@ retry.attempts = 3
+[alembic]
+# path to migration scripts
+script_location = {{ cookiecutter.repo_name }}/alembic
+file_template = %%(year)d_%%(month).2d_%%(day).2d_%%(hour).2d_%%(minute).2d_%%(second).2d_%%(rev)s
```
I'm not sure we should change the default file_template versus stock alembic

hyperknot replied:
I vote for the date based file template, otherwise it's super hard to find the latest revision.

stevepiercy replied:
@mmerickel what are you unsure about? Can you elaborate?

@hyperknot does Alembic then support using the custom identifiers as [partial identifiers](http://alembic.zzzcomputing.com/en/latest/tutorial.html#partial-revision-identifiers), e.g., `alembic upgrade 20180608141155`?

Would we lose any feature from Alembic with custom identifiers?

hyperknot replied:
I've just checked, alembic always works with the revision hash, no matter what the filename is. 
So the command will not change, it'll still be `alembic upgrade e2845a63341f` (or the first few characters).
The date part is for the developers to see and manage revisions in a logical order.

mmerickel replied:
I'd propose the following file template if we're going to change it.

```
file_template = %%(year)d%%(month).2d%%(day).2d_%%(rev)s
```

It's simple and sortable without too much information in it. Compare `20180611_25c1450def88.py` to your proposed `2018_06_11_12_02_37_25c1450def88.py`.

stevepiercy replied:
I like that because there is a single underscore that allows me to easily parse the revision for `alembic upgrade 25c`. If I need further sorting for a given date, I can look at file timestamps.

hyperknot replied:
Timestamps disappear in git, but overall I guess the shorter version is a good compromise.

mmerickel commented on {{cookiecutter.repo_name}}/README.txt:
```diff
@@ -20,9 +20,28 @@ Getting Started
+    - Later you can run database migrations.
+    - Also there is commented code that can be used to
+    - initialize the database directly to the latest revision.
+    - {{cookiecutter.repo_name}}/scripts/initializedb.py
```
I do not like leaving commented code in here for an alternate workflow. There should either be 1) an option in the cookiecutter.json to select with or without alembic or 2) we update the cookiecutter to only support the alembic workflow.

I prefer option 2.

stevepiercy replied:
Without migrations, tutorial users are left to figure out on their own how to update their database after initialization. Users may follow the example in the tutorial and continually update the `initializedb.py` script only until they finally realize it's a bad idea. I agree with @mmerickel to support only the Alembic workflow.

ibeex replied:
So there should be only option 2) alembic way with migrations. Is this final opinion?

stevepiercy replied:
@ibeex yes, please. That is the consensus.

ibeex replied:
I am hawing trouble with initialize_db.py
after removing (b) option from readme
```
- Initialize the database.

    - This cookiecutter provides support for initializing the database
    - with migrations through alembic
    
    - Initialize the database with migrations through alembic:
    - First use alembic to generate revisions.

        env/bin/alembic -c development.ini revision --autogenerate -m "init"

    - upgrade to that revision:

        env/bin/alembic -c development.ini upgrade head
    
    - Later you can run database migrations.
```
I don't see need for `initialize_db` all db should be handeled with `env/bin/alembic ...` so I would remove `initialize_db.py` from role off creating db and leave it only as way to insert first row. That way user can folow http://alembic.zzzcomputing.com/en/latest/tutorial.html#running-our-second-migration doc's for later workflow.

mmerickel replied:
I think this approach is ok with me for handling schema changes. In my projects I usually have two commands `db:migrate` and `db:load-fixtures`. Performing an `alembic upgrade head` is equivalent to `db:migrate` whereas right now we do not have anything for loading fixtures (stuff like default data in tables) - a pattern I think some people have been putting into the `initialize_db` script. This is a good opportunity to add that missing pattern but I'm ok waiting until after this is merged to do that if we wish.

mmerickel reviewed (changes requested):
There are 2 workflows.

1) Initialize a new database.
2) Migrate an existing database.

Currently we say for 1 that you should run `initialize_db` and for 2 you should run `alembic upgrade head`. I would prefer that we standardize this into a single script that does both. A typical script I use for this that doesn't exactly match what we want but is very close is this and I propose we do something like that here and keep calling it `initialize_db`:

```python
from alembic.command import stamp, upgrade
from alembic.config import Config

from myapp.model.meta import metadata
from myapp.model.meta import get_current_version, get_engine

log = __import__('logging').getLogger(__name__)

def main(cli, args):
    cfg = Config(args.config_file)
    engine = get_engine(cli.settings, check_version=False)
    version = get_current_version(engine)

    if version is None:
        log.debug('initializing the database')
        metadata.create_all(bind=engine)
        stamp(cfg, 'head')

    else:
        log.debug('executing migrations')
        upgrade(cfg, args.rev)

    version = get_current_version(engine)
    log.info('database is now at revision=%s', version)
```

stevepiercy commented:
One tiny nit: the cookiecutter currently uses `initializedb` instead of `initialize_db`. I think we should use `initialize_db` instead not only for legibility, but also because that is the name which the [original alchemy scaffold uses](https://github.com/Pylons/pyramid/blob/master/pyramid/scaffolds/alchemy/setup.py_tmpl#L53).

It would be nice to correct this oversight in this PR while we're at it, since I'm going to need to rewrite docs for it anyway.

ibeex committed:
only alembic way

ibeex commented:
Here it is, removed option without migrations, changed revision template, and some error messages to reflect new flow.

stevepiercy committed:
Reduce verbosity and redundancy, correct indentation

stevepiercy committed:
Make consistent with README.rst

stevepiercy committed:
Make consistent with README.rst

hyperknot commented:
One thing related to this PR, what I've discussed with @mmerickel a few month ago on IRC is a template for utility scripts. That is command line scripts in cron jobs, etc.

Originally I was basing my scripts on initialize_db, but I had to rewrite functions not to use the request object. The solution was that the recommended way is to use a request object so that the same functions can be used as in the internet facing code. This means transaction manager should also come from request, not from import transaction. I also had some hard to find bugs when I was using `transaction.manager`, which disappeared when I changed to `request.tm`.

My skeleton for utility scripts as of today is now the following. Maybe it'd be nice to have this in the PR, as I guess command line scripts are very common in web apps.

```
from pyramid.paster import get_appsettings, setup_logging, bootstrap
from pyramid.request import Request


def do_something():
    with request.tm:
        user_ids = [u.id for u in request.dbsession.query(User)]

    # something possibly slow, without locking transactions
    for i, user_id in enumerate(user_ids):
        with request.tm:
            user = request.dbsession.query(User).get(user_id)
            # do something ...


if __name__ == '__main__':
    # init
    setup_logging(config_uri)
    settings = get_appsettings(config_uri)

    base_url = ...... 

    # request
    request_dummy = Request.blank('/', base_url=base_url)
    env = bootstrap(config_uri, request=request_dummy)
    request = env['request']

    do_something()

    env['closer']()
```

mmerickel commented:
@hyperknot Yes, I think we should discuss the utility script in a separate PR after alembic support is merged. The specific example would be a script to load fixtures to replace the functionality we're losing by dropping the initialize_db script. The current initialize_db script could never be idiomatic as a typical script because it's in the unique position of working with an uninitialized database so it has to do things a little differently.

hyperknot commented:
@mmerickel nice, I agree with that approach.

stevepiercy commented:
@mmerickel are you waiting for @ibeex to [make the changes according to your comment](https://github.com/Pylons/pyramid-cookiecutter-alchemy/pull/7#pullrequestreview-127284028)? I think something got lost in communication here.

mmerickel committed:
minor tweaks to reuse existing code

mmerickel commented:
@ibeex I've opened https://github.com/ibeex/pyramid-cookiecutter-alchemy/pull/1 with some extra changes. Could you please review?

mmerickel committed:
update initialize_db to use bootstrap

mmerickel committed:
document running initialize_db

mmerickel commented:
My only other comment is that it's an open debate whether we recommend using `bootstrap` versus the pattern in `initialize_db` in this PR. I much prefer the current approach without bootstrap for performance (it doesn't import your pyramid config) and is what I use in my own projects because I don't see any need for my CLI scripts to import / depend on pyramid. However, the pattern using bootstrap is much simpler and might be better for the purposes of the cookiecutter.

Right now I'm leaning toward the bootstrap approach. I have pushed a second PR that includes bootstrap to review the simplicity. See #33 for a PR to my edit branch that makes this change.

hyperknot commented:
I do find #33 to be cleaner. (And how much does a fraction of a second matter in a script which is run exactly once per project?)

mmerickel commented:
Sure, the main downside is that as your app grows you may put more web-specific things into your wsgi `main(global_config, **settings)` which you may not want to run in every script - but I think that bootstrap is a great start.

hyperknot commented:
But even in a full blown webapp, we are talking about a couple of seconds, aren't we? And CLI scripts are run once per day or maybe once per hour, I believe.

ibeex committed:
Merge pull request #1 from Pylons/alembic_support-edits

minor tweaks to reuse existing code

mmerickel commented:
@ibeex I had to open https://github.com/ibeex/pyramid-cookiecutter-alchemy/pull/2 since you merged the other PR before https://github.com/Pylons/pyramid-cookiecutter-alchemy/pull/33 was merged into it.

ibeex committed:
Merge pull request #2 from Pylons/use-bootstrap

update initialize_db to use bootstrap

ibeex commented:
@mmerickel merged

mmerickel commented:
great work @ibeex and thanks for the input from everyone - this is exciting to get in and now we need to update the wiki2 tutorial a bit to explain the migration setup in https://docs.pylonsproject.org/projects/pyramid/en/1.9-branch/tutorials/wiki2/basiclayout.html

stevepiercy commented:
I've started work in a PR https://github.com/Pylons/pyramid/pull/3307 on a new branch `alembic-cookiecutter-alchemy` in Pyramid.
